### PR TITLE
[auto-parallel] refactoring ColoTracer

### PR DESCRIPTION
**Pull request by superhao1995**

### CODE REVIEW

File: colossalai/fx/tracer/experimental.py [hunk 1]
```diff
@@ -392,3 +388,77 @@ def wrapper(*args, **kwargs):
     def __exit__(self, exc_type, exc_val, exc_tb):
         for name, (wrapper, orig) in self.overrides.items():
             setattr(torch, name, orig)
+
+
+def set_meta_data_for_node(gm: ColoGraphModule, root: torch.nn.Module, meta_args: Optional[Dict[str, Any]]=None):
+    for node in gm.graph.nodes:
+        node._meta_data = _meta_data_computing(meta_args, root, node.op, node.target, node.args, node.kwargs)
```
Comment (super-dainiu) [inline colossalai/fx/tracer/experimental.py:395]: A better option is to put `._meta_data` into `node.meta['meta_data']`.

File: colossalai/fx/tracer/experimental.py [hunk 2]
```diff
@@ -392,3 +435,151 @@ def wrapper(*args, **kwargs):
...
+    return meta_out
+
+
+def bias_addition_pass(gm: ColoGraphModule, root_model: torch.nn.Module, meta_args: Optional[Dict[str, Any]]=None):
+    meta_prop_pass(gm, root_model, meta_args)
+    result_graph = Graph()
+    value_remap = {}
+    unwrap_fn = lambda n: n._meta_data if isinstance(n, Node) else n
+
+    for ori_node in gm.graph.nodes:
```
Comment (super-dainiu) [inline colossalai/fx/tracer/experimental.py:520]: Let it be `orig_node`. This sounds more **PyTorch**

File: colossalai/fx/tracer/experimental.py [hunk 3]
```diff
@@ -508,3 +509,77 @@ def _meta_data_computing(meta_args, root, kind, target, args, kwargs):
         raise RuntimeError(f"Could not compute metadata for {kind} target {target}: {e}")
 
     return meta_out
+
+
+def bias_addition_pass(gm: ColoGraphModule, root_model: torch.nn.Module, meta_args: Optional[Dict[str, Any]]=None):
+    meta_prop_pass(gm, root_model, meta_args)
```
[the remark this answers is not in the document]
Comment (superhao1995) [reply colossalai/fx/tracer/experimental.py:514]: ok

File: colossalai/fx/tracer/experimental.py (hunk 2 above)
Comment (superhao1995) [reply colossalai/fx/tracer/experimental.py:519]: ok

File: colossalai/fx/tracer/experimental.py [hunk 4]
```diff
@@ -392,3 +441,202 @@ def wrapper(*args, **kwargs):
...
+    sig_names = set(sig.parameters.keys())
+    meta_arg_names = set(meta_args.keys())
+
+    # update concrete args with default values
+    non_meta_arg_names = sig_names - meta_arg_names
+    for k, v in sig.parameters.items():
+        if k in non_meta_arg_names and \
+                k not in concrete_args and \
+                v.default is not inspect.Parameter.empty:
+            concrete_args[k] = v.default
```
Comment (super-dainiu) [inline colossalai/fx/tracer/experimental.py:468]: this part seems to be a replica of the previous one?

File: colossalai/fx/tracer/experimental.py [hunk 5]
```diff
@@ -263,11 +281,42 @@ def _check_arg_name_valid(names):
...
+                    raise NotImplementedError(
+                        "We do not implement the backward pass as we only trace the forward pass.")
+
+            # override the checkpoint function
+            torch.utils.checkpoint.CheckpointFunction = PatchedCheckpointFunction
+        yield
+
+        if enabled:
+            # recover the checkpoint function upon exit
+            torch.utils.checkpoint.CheckpointFunction = orig_ckpt_func
```
Comment (super-dainiu) [inline colossalai/fx/tracer/experimental.py:317]: u may try to generalize this to a recursive checkpoint function. shud be easy to implement.

File: colossalai/fx/tracer/experimental.py [hunk 6]
```diff
@@ -263,11 +281,42 @@ def _check_arg_name_valid(names):
...
+    def trace_activation_checkpoint(self, enabled: bool):
+        if enabled:
+            orig_ckpt_func = torch.utils.checkpoint.CheckpointFunction
+
+            class PatchedCheckpointFunction(torch.autograd.Function):
+
+                @staticmethod
+                def forward(ctx, run_function, preserve_rng_state, *args):
+                    # signal that the current tracing occurs within activaton checkpoint part
+                    self.inside_torch_checkpoint_func = True
```
Comment (super-dainiu) [inline colossalai/fx/tracer/experimental.py:300]: now this flag should be a fifo list.
sth like [outer_ckpt_region, inner_ckpt_region, ...]
before entering a ckpt func, push the region idx
after exiting, pop the region idx

**Review by YuliangLiu0306 — APPROVED** [review]

**Pull request merged**